[RHCLOUD-18437] Add unique index into applications table for source id + app type id + tenant id

petracihalova opened the pull request:
One source can't have 2 applications with same app type - to prevent this, we need to add a unique index to the db.
**JIRA:** [RHCLOUD-18437](https://issues.redhat.com/browse/RHCLOUD-18437)


But first we need to be sure that in the stage and production environment we don't have this data.
We can merge this migration after [RHCLOUD-19355](https://issues.redhat.com/browse/RHCLOUD-19355).

Commit:
added migration that creates a unique index for app table for source id, app type id and tenant id

MikelAlejoBR commented on db/migrations/20220510112000_add_application_constraint.go:
```diff
@@ -0,0 +1,39 @@
+		gorm.Model
+		SourceID          int64 `gorm:"uniqueIndex:index_applications_on_source_id_tenant_id_app_type_id"`
+		ApplicationTypeID int64 `gorm:"uniqueIndex:index_applications_on_source_id_tenant_id_app_type_id"`
+		TenantID          int64 `gorm:"uniqueIndex:index_applications_on_source_id_tenant_id_app_type_id"`
```
I would say to stick to how [Postgres names its indexes](https://stackoverflow.com/a/4108266). So that would make the index name `applications_app_type_id_source_id_tenant_id_idx`.

petracihalova replied:
well I followed what we already have in db ... e.g. these are our existing indexes names in applications table 
![image](https://user-images.githubusercontent.com/89980168/168595238-75e16454-77a5-43a5-b0eb-3988397a5eba.png)
 
My first option was `"idx_xxxxx"` because that is how it is used in [documentation](https://gorm.io/docs/indexes.html), but then I decided to let it in same way how we have other existing indexes

MikelAlejoBR replied:
I see, but there's a couple of things:

* We still have that naming because the PR which transforms all those names is still pending to be merged: https://github.com/RedHatInsights/sources-api-go/pull/175
* I chose Postgres naming because that's standard to any user that uses Postgres, instead of framework specific like it was with Active Record, or like it is with Gorm.

Anyway we can merge as it is, and I'll update my PR afterwards.

petracihalova replied:
I see 😁  thanks for the reminder, i completely forgot we have these pending PRs ... I will set same naming here 👍🏻

MikelAlejoBR reviewed (approved):
LGTM!

Commit:
updated app fixtures to have there unique combinations of tenant id, app type id and source id

Commit:
fix TestApplicationCreateGood() to create new combination of app type and source id for given tenant

Commit:
updated TestDeleteApplication() to better create test data

bacause now we have unique index in applications table to have unique
combinations of tenant id, source id and application type id, we needed
to create unique application for our test

Commit:
updated TestApplicationDeleteCascade() to better create test data

because now we have unique index in applications table to have unique
combinations of tenant id, source id and application type id, we needed
to create unique application for our test

Commit:
deleted TestApplicationAuthenticationsByApplicationsDatabase() because it violates the unique index in application table

all created test data were in conflict with new unique index in application table (tenant id + application type id + source id), the test will be replaced with new one in different commit

Commit:
added new test for the method ApplicationAuthenticationsByResource() and resource type = 'Source'

Commit:
splitted TestDeleteCascade() into two separate tests

one test contains the delete test for source with sub resources
(apps, auths, endpoints ...) and second test contains the same for
source without subresources

Commit:
refactor of TestDeleteCascade() to create only 1 set of subresources for the source

because of new unique index in application table it is no long possible to
create more than 1 application of specific app type for a source and tenant
so this test is updated to create only one set of subresources

lindgrenj6 reviewed (approved):
Dupes have been eliminated on stage/prod.
